### PR TITLE
[improve][client] PIP-224 Part 1: Add TopicMessageId for seek and acknowledge

### PR DESCRIPTION
**Pull request by BewareMyPower**
Master Issue: https://github.com/apache/pulsar/issues/18616

### Motivation

Introduce `TopicMessageId` to support getting the owner topic of a `MessageId`. When a `MessageId` is retrieved from a received message, the owner topic will be correctly set by the client library. When it's returned by `Producer#send`, this PR provides a `TopicMessageId#create` method to configure the owner topic.

`acknowledge` APIs are affected only for the error cases: when a `MessageId` other than a `TopicMessageId` is accepted on a multi-topics consumer, `PulsarClientException.NotAllowedException` will be thrown.

The semantic of the `seek(MessageId)` API is changed. Now if a `TopicMessageId` is accepted on a multi-topics consumer, the seek behavior will happen on the internal consumer of the owner topic.

### Modifications

- Add the `TopicMessageId` interface.
- In `MultiTopicsConsumerImpl#doAcknowledge`, complete the future with `NotAllowedException` if the argument is not a `TopicMessageId`.
- In `MultiTopicsConsumerImpl#seekAsync`, when the argument is a `TopicMessageId`, find the internal consumer according to the owner topic and pass the argument to it if it exists.
- In `ConsumerImpl#seekAsync`, get the inner message ID of the `TopicMessageId` so that now a single-topic consumer can also accept a `TopicMessageId` to seek.

Besides the main modifications above, this patch does some refactorings to avoid direct access to `TopicMessageIdImpl`:
- Deprecated `getTopicName` method by trimming the partition suffix of the owner topic in `getOriginTopicNameStr`.
- Deprecated `getTopicPartitionName` by `getOwnerTopic`.
- `getInnerMessageId` cannot be deprecated because we still need to convert `TopicMessageId` to `MessageIdImpl` in many cases (because we cannot get the fields like ledger id). Instead of deprecating it, use `MessageIdImpl.convertToMessageIdImpl` to replace it.
- In `convertToMessageIdImpl`, for a customized `TopicMessageId` implementation, use serialization and deserialization to get the `MessageIdImpl` object.

### Verifications

Add the following tests to `MultiTopicsConsumerTest`:
- `testAcknowledgeWrongMessageId`: verify the correct exceptions are thrown in `acknowledge` APIs
- `testSeekCustomTopicMessageId`: verify the new seek semantics for a `TopicMessageId`, including the existing `TopicMessageIdImpl` and the customized implementation by `TopicMessageId#create`

### TODO
- Add a standard SerDes class for `TopicMessageId`
- Apply `TopicMessageId` into `getLastMessageId` related APIs.
- Deprecate the `getInnerMessageId` after PIP-229 is approved.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/BewareMyPower/pulsar/pull/17

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java [hunk 1]
```diff
@@ -49,6 +49,8 @@ public interface MessageAcknowledger {
      *
      * @throws PulsarClientException.AlreadyClosedException}
      *             if the consumer was already closed
+     * @throws PulsarClientException.NotAllowedException
+     *             if `messageId` is not a {@link TopicMessageId} when multiple topics are subscribed
```
Comment (codelipenghui) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java:53]: And is it also for the partitioned topic? Users need to use 

```java
MessageId.fromByteArrayWithTopic();
```

or 

```java
TopicMessageId.create(topic, MessageId.fromByteArray());
```

Does the TopicMessageId interface is intentionally exposed to users? Which one should be suggested in the future?

Comment (BewareMyPower) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java:53]: > Does the TopicMessageId interface is intentionally exposed to users? 

Yes. The new `getLastTopicMessageId` APIs will return `TopicMessageId` instances, they will be introduced in another PR, see the [original PIP issue](https://github.com/apache/pulsar/issues/18616).. And users can avoid touching the `pulsar-client` module (internal implementations) to get the topic name of a message ID.

```java
if (msgId instanceof TopicMessageId) {
    multiTopicsConsumer.seek(msgId);
} else {
    multiTopicsConsumer.seek(TopicMessageId.create(topic, msgId));
}
```

> Which one should be suggested in the future?

`fromByteArrayWithTopic` should be deprecated. The `MessageId` that is not a `TopicMessageId` could be retrieved from other ways, not only the deserialization. But `fromByteArrayWithTopic` assumes it (e.g. a `MessageIdImpl`) was retrieved from deserialization.

Comment (BewareMyPower) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java:53]: > Does the TopicMessageId interface is intentionally exposed to users? 

Yes. The new `getLastTopicMessageId` APIs will return `TopicMessageId` instances, they will be introduced in another PR, see the [original PIP issue](https://github.com/apache/pulsar/issues/18616).. And users can avoid touching the `pulsar-client` module (internal implementations) to get the topic name of a message ID.

```java
if (msgId instanceof TopicMessageId) {
    multiTopicsConsumer.seek(msgId);
} else {
    multiTopicsConsumer.seek(TopicMessageId.create(topic, msgId));
}
```

> Which one should be suggested in the future?

`fromByteArrayWithTopic` should be deprecated. The `MessageId` that is not a `TopicMessageId` could be retrieved from other ways, not only the deserialization. But `fromByteArrayWithTopic` assumes it (e.g. a `MessageIdImpl`) was retrieved from deserialization.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+public interface TopicMessageId extends MessageId {
+
+    /**
+     * Return the owner topic name of a message.
+     *
+     * @return the owner topic
+     */
+    String getOwnerTopic();
+
+    static TopicMessageId create(String topic, MessageId messageId) {
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java:42]: I changed it to an inner class.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/MessageAcknowledger.java:53]: Ok, got it.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java:42]: Looks like you haven't pushed the commit?

[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TopicMessageId.java:42]: Oh, sorry, I pushed it now.

**Review by Technoboy- — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**